Text on proof automation

goens opened the pull request:
See #920 

The current structure (and even title modification) are just suggestions, happy to discuss them in this PR of course!

Closes #920

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
-\note{TODO: expand this sketch}
+In this project we used proof automation in two ways: automated theorem provers (ATPs) and Lean tactics.
+ATPs are generally stand-alone tools that implement a (semi-) decision procedure for a given formal language or related set of languages.
+For example, Vampire~\cite{DBLP:conf/cav/KovacsV13} is an ATP focused primarily on first-order logic using supperposition, which we used extensively in this project.
```
supper->super

pitmonticone replied:
Fixed. Thanks.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+ATPs are generally stand-alone tools that implement a (semi-) decision procedure for a given formal language or related set of languages.
+For example, Vampire~\cite{DBLP:conf/cav/KovacsV13} is an ATP focused primarily on first-order logic using supperposition, which we used extensively in this project.
+ATPs are generally complex pieces of software, and can often have bugs.
+While we could choose to trust the results of an ATP, and in this project we don't do so either.
```
Needs rewrite.

vlad902 replied:
I would merge with the previous/next sentence and make it more explicit: "ATPs are complex software that can contain bugs. Instead of trusting ATP output, we used proof certificates, which many ATPs can produce, to reconstruct a proof in Lean.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+ATPs are generally complex pieces of software, and can often have bugs.
+While we could choose to trust the results of an ATP, and in this project we don't do so either.
+Instead, we use proof certificates, which many ATPs can produce, to reconstruct a proof in Lean.
+This process depends on the proof certificate and the ATP, and we describe it for the main recunstruction we've done.
```
recunstruction -> reconstruction. we've -> we have.

pitmonticone replied:
Fixed. Thanks.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+In other words, tactics are programs that operate at the meta-level of Lean code: they essentially take in Lean code as input and produce Lean code as output.
+In this manner, they look like another keyword in the language, and are tightly integrated by producing proofs directly.
+Under the hood, they can implement essentially anything, from syntactic sugar to full decision procedures.
+The \texttt{duper} tactic~\cite{DBLP:conf/itp/CluneQBA24}, for example, implements a superposition calculus, similar Vampire's, but for dependent types --- Lean's underlying logical foundation.
```
similar to

pitmonticone replied:
Fixed. Thanks.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+
+\paragraph{Saturation}
+Most of the ATPs we used extensively in this project rely primarily on saturation procdures in the superposition calculus.
+This is the case for Vampire~\cite{DBLP:conf/cav/KovacsV13}, and the paper alo gives a more in-depth explanation of this.
```
alo -> also. though i'd probably rephrase this to use the citation as the object in that clause.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+See also~\cite{DBLP:journals/cacm/BentkampBNTVW23} for a gentler exposition.
+The core idea of these provers is that they take a set of assumptions and conjecture, expressed in --- say --- first-order logic.
+They take the conjecture and negate it, adding this negation to the set of assumptions, which are all put in some normal form.
+The ATP then try to refute the negation by applying rules of the underlying calculus, until they find a proof of false (a contradiction).
```
try -> tries

pitmonticone replied:
Fixed. Thanks.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+This can also be read as $C \lor L \quad D \lor \neg L$ implies $C \lor D$, where $C, D, L$ are formulas in e.g. first-order logic.
+Superposition calculi have a variant of this rule that deals with equality directly, and thus are more efficient at reasoning about equality.
+
+In this project we used Vampire~\cite{DBLP:conf/cav/KovacsV13}, Duper~\cite{DBLP:conf/itp/CluneQBA24} and Prover9 and Mace4~\cite{prover9-mace4} which are all based on variants of saturation for proving.
```
Does mace4 count as a saturation-based prover? I'm not an expert here.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+\subsection{Proof Reconstruction and Integration}
+\label{sec:proof-reconstruction}
+
+While ATPs are very useful to solve questions for this project, they generally don't integrate well with Lean (except for Lean tactics like \texttt{duper}, \texttt{aesop} and \texttt{egg}).
```
"While ATPs are very useful to solve questions for this project, they generally don't integrate well with Lean" I would be more specific: "ATPs can resolve many implications or non-implications studied by this project; however, except for Lean tactics like ..., ATP systems do not generate proof certificates that are natively recognized by Lean."

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+This project primarily used the least integrated approach, (Option \ref{external}), as it was fastest and required no dependencies on the other contributors.
+This also has drawbacks: primarily, while the upfront effort is lower, the effort to use is higher than with a tactic, once the tactic is developed.
+It also makes integration with larger code bases more difficult: in this project the (mathematical) dependencies were by design very minimal, Magmas are simple, and we built our own definitions for them.
+For example, integration with the typeclass system become much more difficult when working with more complex mathematical objects that build upon many layers of structure.
```
become -> becomes. I don't understand what this is conveying

pitmonticone replied:
Fix the typo, but I will not resolve the conversation because of the comment.

goens replied:
I don't really understand what you don't understand about it. I expanded it more with a citation, hopefully that helps.

vlad902 replied:
After re-reading I'm not sure what my original confusion was about.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+
+This project primarily used the least integrated approach, (Option \ref{external}), as it was fastest and required no dependencies on the other contributors.
+This also has drawbacks: primarily, while the upfront effort is lower, the effort to use is higher than with a tactic, once the tactic is developed.
+It also makes integration with larger code bases more difficult: in this project the (mathematical) dependencies were by design very minimal, Magmas are simple, and we built our own definitions for them.
```
Magmas -> magmas

pitmonticone replied:
Fixed. Thanks.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+Another use of ATPs has been in a semi-automatic fashion, to find counter-examples.
+The general strategy was to use ATPs to find counter-examples to implications by building magmas iteratively.
+If we want to build a counterexample to $l \vdash l'$, we want to build construct a magma where $l$ holds but $l'$ doesn't.
+In this method, we iteratively strengthen a construction with additional hypotheses, and use the ATP to check wether these hypotheses are not too strong (to imply $l'$) or unsound (to disallow $l$).
```
wether -> whether

pitmonticone replied:
Fixed. Thanks.

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+
+\paragraph{Semi-Automated Counterexample Guidance}
+
+Another use of ATPs has been in a semi-automatic fashion, to find counter-examples.
```
I would maybe get rid of "in a semi-automatic fashion," here. There are examples of fully-automated and quite bespoke counterexample construction. It may be interesting to mention that, or leave a TODO for the future.

goens replied:
I don't understand the issue with that, but I have also been less involved in this effort. Semi-automated subsumes fully automated, and not all have been fully automated, right? More importantly though, the entire section has talked about fully automated approaches, this is just saying it's not *only* been fully automated, there have been semi-automated approaches. Maybe this should be expanded in a different PR after?

vlad902 commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+
+TODO: this should also be expanded more, at least with references to some of the constructions in other chapters.
+
+While equational theories can also be used in a semi-automatic fashion to prove equations~\cite{DBLP:journals/pacmpl/KoehlerGBGTS24}, the positive implications in this project were all simple enough that we did not need a semi-automatic approach for them.
```
equational theories -> equality saturation? While this sentence is true for the general graph, it is not completely true for either the finite implications or the Higman-Neumann work Jose Brox has done, where he has indeed performed guided search.

goens replied:
fair point! I changed the formulation a bit and added a TODO to include those.

teorth commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+The core idea of these provers is that they take a set of assumptions and conjecture, expressed in --- say --- first-order logic.
+They take the conjecture and negate it, adding this negation to the set of assumptions, which are all put in some normal form.
+The ATP then try to refute the negation by applying rules of the underlying calculus, until they find a proof of false (a contradiction).
+In this case, the conjecture was (classicaly) true, and the ATP has found a proof by contradiction, often called a ``refutation'' or ``saturation'' proof.
```
"classicaly" -> "classically"

teorth commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+One of the main advantages of using equational reasoning to reason about implications of magma laws is that we get very explicit proofs: a proof that $l \vdash l'$ is given by a sequence of rewrites that starts at the left-hand side of $l'$ and arrives at the right-hand side through applications of $l$.
+
+In this project we used Z3~\cite{DBLP:conf/tacas/MouraB08}, Prover9 and Mace4~\cite{prover9-mace4}, a custom ATP for magmas based on egg~\cite{DBLP:journals/pacmpl/WillseyNWFTP21}, and the Lean egg tactic~\cite{DBLP:journals/pacmpl/KoehlerGBGTS24,rossel2024bridging}, which all work with equational logic. We have also reasoned with manual (custom written) heuristics about simple rewrites.
+
```
It would be great to have some illustrations (perhaps as figures or screenshots) of some sample outputs (e.g. a command line interface input and output) of some of these ATP tools in action for a sample implication in the project.  Perhaps one can add one such example somewhere around here and make a stub to ask for other contributions of this type?  Ideally one should have examples from two or three different representative ATPs.

teorth commented on paper/automated.tex:
```diff
@@ -1,20 +1,108 @@
+For equational proofs from external provers (e.g. MagmaEgg), we used also a simple version of reconstruction by (re-)constructing the proofs of equality from an explanation, using congruence lemmas from Lean.
+In the case of equational proofs from the \texttt{egg} tactic, these could be converted into a series of \texttt{calc} steps, documenting the explicit calculation, using the \texttt{calcify} tactic\footnote{\url{https://github.com/nomeata/lean-calcify/}}.
+
+In general, we've observed that there are multiple ways of integrating decisions procedures within Lean, with different levels of integration.
```
I would generally avoid contractions such as "we've" in academic writing.

pitmonticone commented:
I have directly fixed a bunch of typos I have found together with all those found by @vlad902 and resolved the relevant review comments.

goens commented:
I think I've addressed all the suggestions now. Happy to have either a second round of reviews/suggestions or merge this and continue expanding with more PRs.

teorth commented:
I think it's good to merge for now.  If anyone wants to work on the section more, they can say so in the Zulip chat, and I'll open a new issue for it.
